x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/StringMatcher.java

```diff
@@ -118,6 +118,9 @@ public StringMatcher build() {
             }
 
             final String description = describe(allText);
+            if (nonExactMatch.contains("*")) {
+                return new StringMatcher(description, s -> true);
+            }
```

Comment:
Since we are trying to squeeze every bit for performance, I wonder whether it could help a tiny bit if we perform this check in the `include(String pattern)` method. It already checks for wildcard and should be easy to tweak it to flag for seeing a match-all pattern.

Reply:
That was my original implementation, but it's really not necessary.
We're not looking to squeeze that level of performance in the Builder, just in the StringMatcher itself.
Testing every pattern that is included to see whether it's equal to `*` and then setting a flag isn't really going to make a difference, and it means carrying more state in the builder.